distribution_metadata.c

```diff
@@ -389,7 +391,7 @@ IsDistributedTable(Oid tableId)
 	 * The query below hits the partition metadata table, so if we don't detect
 	 * that and short-circuit, we'll get infinite recursion in the planner.
 	 */
-	if (tableId == partitionMetadataTableOid)
+	if (IsSystemNamespace(tableNamespaceOid) || tableId == partitionMetadataTableOid)
```

Comment:
Well, they're both queries against the relation named `pgs_distribution_metadata.partition`. That relation is either a view into the CitusDB tables or a normal table. In the view case, the query gets rewritten into a subquery like `SELECT NULL FROM (SELECT logicalrelid AS relation_id ... FROM pg_dist_partition) WHERE relation_id = $1`. Our planner sees the reference to `pg_dist_partition` in the subquery and we hit recursion there without this check.

IMO it's still probably more consistent to just check whether the table is in `pgs_distribution_metadata` or in a system catalog, but in an earlier comment you asked me to check for the partition table `oid` directly. But since we now also have to handle a CitusDB catalog, and since I don't want to have the C code know about the `oid`s of CitusDB tables, I check for whether the target table is a catalog table (which works in CitusDB).

As for the comment; clearly it needs some clarification :grin:. Updating. (done)